validation/linux_cgroups_devices.go

```diff
@@ -0,0 +1,22 @@
+package main
+
+import (
+	"github.com/opencontainers/runtime-tools/cgroups"
+	"github.com/opencontainers/runtime-tools/validation/util"
+)
+
+func main() {
+	var major1, minor1, major2, minor2, major3, minor3 int64 = 10, 229, 8, 0, 10, 200
+	g, err := util.GetDefaultGenerator()
+	if err != nil {
+		util.Fatal(err)
+	}
+	g.SetLinuxCgroupsPath(cgroups.AbsCgroupPath)
+	g.AddLinuxResourcesDevice(true, "c", &major1, &minor1, "rwm")
+	g.AddLinuxResourcesDevice(false, "b", &major2, &minor2, "rw")
+	g.AddLinuxResourcesDevice(true, "a", &major3, &minor3, "r")
```

Comment:
@q384566678 I've just looked into this PR again.
True, as you said, when running manually `runc run mycontainer`, the device cgroups list shows the full result correctly.

Apart from that, I think the line above `g.AddLinuxResourcesDevice(true, "a", &major3, &minor3, "r")` causes the issue of the wrong device list.
Think about a sequence of the following command lines.

```
# echo "c 10:229 rwm" > /sys/fs/cgroup/devices/cgrouptest/devices.allow
# cat /sys/fs/cgroup/devices/cgrouptest/devices.list   # This shows a correct result
c 10:229 rwm

# echo "a 10:200 r" > /sys/fs/cgroup/devices/cgrouptest/devices.allow
# cat /sys/fs/cgroup/devices/cgrouptest/devices.list   # an unexpected result. Existing devices are now gone
a *:* rwm
```

According to the [Kernel cgroup v1 document](https://www.mjmwired.net/kernel/Documentation/cgroup-v1/devices.txt#18), doing `echo a > /sys/fs/cgroup/1/devices.allow` will add the 'a *:* rwm' entry to the whitelist. Apparently `allowing all devices` results in wiping out existing entries and adding a single wildcard entry, even when the input is given by a specific pair of major/minor number.

When testing without the line for `a`, it shows a lot better result, doesn't it?

Reply:
> ```
> # echo "a 10:200 r" > /sys/fs/cgroup/devices/cgrouptest/devices.allow
> # cat /sys/fs/cgroup/devices/cgrouptest/devices.list # an unexpected result. Existing devices are now gone
> a *:* rwm
> ```

That's surprising to me.  In the wild, I expect few users to care about major/minor but not block/char.  Still, does anyone have time to file a kernel patch to either respect the passed value or error out when major or minor are passed with `a`?

Reply:
updated, @dongsupark @liangchenye PTAL